man/mkstemp.man
.TH MKSTEMP %manext1% "March 2, 2014" "%nmhversion%"
.\"
.\" %nmhwarning%
.\"
.SH NAME
mkstemp \- create a temporary file
.SH SYNOPSIS
.HP 5
.na
.B %libdir%/mkstemp
.RB [ \-directory
.IR directory ]
.RB [ \-prefix
.IR prefix ]
.RB [ \-suffix
.IR suffix ]
.RB [ \-version ]
.RB [ \-help ]
.ad
.SH DESCRIPTION
.B mkstemp
creates a temporary file using
.IR mkstemp (3),
or if the
.B \-suffix
switch is available and used,
.IR mkstemps (3).
The file name identifies a unique, newly created file, and always
contains 6 characters that appear to be random.
.PP
The
.B \-directory
switch specifies a
.I directory
in which to create the file.
.PP
The
.B \-prefix
switch specifies an initial part, before the 6 characters, of the file
name.
.PP
The
.B \-suffix
switch is only supported on platforms that provide the
.IR mkstemps (3)
library function.  It specifies a
.I suffix
to appear after the 6 characters; if it should begin with a special
character such as a period, that must be explicit in
.IR suffix .
.PP
Unlike arguments to most switches of other
.B nmh
commands, the arguments to
.B mkstemp
switches can begin with a dash.
.SH "EXIT STATUS"
If
.B mkstemp
successfully creates the temporary file, it prints the file name on
standard output and exits with status 0.  On failure, it prints a
diagnostic message on standard error and exits with status -1.
.SH "PROFILE COMPONENTS"
None
.SH "SEE ALSO"
.IR mkstemp (3),
.IR mkstemps (3)
.SH DEFAULTS
.PD 0
.TP 14
\-directory
current directory
.TP
\-prefix
none
.TP
\-suffix
none
.PD


NEWS
#
# NEWS - Release notes for nmh 1.5
#

Welcome to nmh, the new version of the classic MH mail handling
system.  Long-time MH and nmh users should read carefully the NOTABLE
CHANGES section, as there are some subtle but significant changes
to the way nmh now behaves.  Otherwise, please see the README and INSTALL
files for help on getting started with nmh.

------------
NEW FEATURES
------------

The following are new features for the 1.5 release of nmh.

- Improved performance for TLS encryption of the SMTP protocol.
- MH-E and POP support are now always compiled in.
- The selection of a locking directory for dot-locking files is now
  controllable via configure.
- fileproc and mhlproc mh-profile are now obeyed by send, rcvdist,
  and whatnow.
- New mh-format instructions have been implemented: %(putlit),
  %(concataddr), %(myhost), %(myname), %(localmbox).  See mh-format(5)
  for more details.
- Nmh's idea of the local mailbox is now configurable via a new profile
  entry, Local-Mailbox.  See mh-profile(5) for more details.
- comp, forw, and dist now process all component templates through
  mh-format(5), and now all support new command line arguments -from,
  -to, -cc, -fcc, and -subject.
- burst(1) now supports any valid message delimiter according to RFC 934.
- post(8) now requires a From: header in all drafts that it processes;
  see NOTABLE CHANGES below for more details.
- post(8) can now explicitly control the SMTP envelope address either via
  the Sender: header or the new Envelope-From: header.
- The -attach option is now the default for send (using the header
  Nmh-Attachment) and -attachformat 1 is the default attachment type.
- repl and dist now support new switches -atfile and -noatfile to control
  the creation of the ./@ link to the message being replied or
  redistributed.  The current default is -atfile.
- Support (if available) for readline editing and filename completion at
  the WhatNow? prompt.
- Support for using an external command to filter the body of a message
  through in mhl.  See mhl(1) for more details, specifically, the
  "format" keyword, the "formatproc" entry in mh-profile(5), and the
  -fmtproc switch for mhl and repl.
- Preliminary support for improved MIME handling when replying to messages!
  Yes, a long requested feature has a solution.  A perl script
  called replyfilter is available; it is designed to act as a mhl
  external filter to process MIME messages in a more logical way.
  It is available in $(srcdir)/docs/contrib/replyfilter or is
  typically installed as $(prefix)/share/doc/nmh/contrib/replyfilter.
  See the comments at the top of replyfilter for usage information;
  it will likely require some adjustment for your site.  replyfilter
  requires the MIME-Tools and MailTools perl modules.

----------------------------
OBSOLETE/DEPRECATED FEATURES
----------------------------

The following features have either been removed completely or marked
as officially deprecated for removal in a future release of nmh.

- The creation of the ./@ link to the message being replied or
  redistributed is deprecated; -noatfile will be the default in the
  next release.  If there are no requests to maintain -atfile it
  will be removed in a future release.
- Old code enabled by the UCI preprocessor definition has been removed.
- All configuration options that were formerly only adjustable by editing
  config.h have either been moved into configure or removed completely.
- Support for the built-in ftp client (used when dealing with external-body
  message parts in MIME messages) has been removed.
- The following environment variables are marked as obsolete/deprecated.
  Support for them will be removed from the next nmh release:
    MHPOPDEBUG (use -snoop command line switch instead)
    MM_NOASK (use -nolist and -nopause command line switches instead)
    NOMHNPROC (use -nocheckmime command line switch instead)
    FACEPROC (undocumented faceproc feature to be removed)
- The hostable option of mts.conf is no longer needed and has been removed.

---------------
NOTABLE CHANGES
---------------

The largest change is that post(8) now requires a From: header in all drafts
that it processes.

The impetus for this change has been the evolution of email and client
systems.  When MH was first developed the norm was to have tightly managed
clients that were part of the same domain as a user's email address.  This
is no longer a valid assumption, and the algorithm used by nmh to determine
the user's local mailbox is no longer a reasonable default.

Unfortunately, the way that nmh uses the user's local mailbox used to happen
very late in the message process; it was handled by post during message
submission.  If no From: header exists in the current draft post would
insert one based on it's idea of the local mailbox, which was frequently
wrong.

After a long discussion within the nmh development community, it was
decided that the only reasonable way forward was to move the local mailbox
handling up a level and place it under more direct control of the end
user.  post(8) no longer will try to determine the local mailbox name;
it will take that information from the draft message.  Thus a From: header
is now required in all drafts.

As part of these changes, all message composition utilities now process
their component files using mh-format.  All of the default component
files have been modified to include a From: header in them; this will
give the user a chance to see nmh's idea of the local mailbox and modify
it if desired.  The new mh-profile(5) entry Local-Mailbox can be used
to adjust the local mailbox name if the system default is not appropriate.

So how should end-users adapt to these changes?  Here are some suggestions:

1) If you use the default component files, everything should basically work
   the same for you.  You should see a From: header in the drafts of your
   messages.  If the address you see in the From: header is incorrect

uip/mkstemp.c
/*
 * mkstemp.c -- create a temporary file
 *
 * This code is Copyright (c) 2014 by the authors of nmh.
 * See the COPYRIGHT file in the root directory of the nmh
 * distribution for complete copyright information.
 */

/* define NMH to 0 to remove dependencies on nmh. */
#ifndef NMH
#   define NMH 1
#endif /* ! NMH */

#ifdef HAVE_CONFIG_H
#   include <config.h>
#endif /* HAVE_CONFIG_H */
#include <unistd.h>
#include <stdlib.h>
#include <string.h>
#include <stdio.h>

#if ! defined HAVE_MKSTEMPS
#   define HAVE_MKSTEMPS 0
#endif /* ! HAVE_MKSTEMPS */

char *build_template(const char *, const char *, const char *);
void process_args(int, char **, const char **, const char **, const char **);

/*
 * Use a template of the form:
 *     [directory/][prefix]XXXXXX[suffix]
 * where some of those named components might be null.  suffix is only
 * supported if HAVE_MKSTEMPS.
 */

int
main(int argc, char *argv[]) {
    const char *directory = "", *prefix = "", *suffix = "";
    size_t suffix_len;
    int fd;
    char *template;

    process_args(argc, argv, &directory, &prefix, &suffix);
    if ((template = build_template(directory, prefix, suffix)) == NULL) {
        return -1;
    }

    if ((suffix_len = strlen(suffix)) > 0) {
#       if HAVE_MKSTEMPS
        if ((fd = mkstemps(template, suffix_len)) < 0) { perror("mkstemps"); }
#       else  /* ! HAVE_MKSTEMPS */
        fd = 0;
#       endif /* ! HAVE_MKSTEMPS */
    } else {
        if ((fd = mkstemp(template)) < 0) { perror("mkstemp"); }
    }

    if (fd >= 0) {
        (void) puts(template);
        (void) close(fd);
    }

    free(template);

    return fd >= 0  ?  0  :  -1;
}


char *
build_template(const char *directory, const char *prefix, const char *suffix) {
    const char pattern[] = "XXXXXX";
    size_t len, directory_len, pathsep_len, prefix_len, suffix_len;
    char *template;

    directory_len = strlen(directory);
    if (directory_len > 0) {
        pathsep_len = 1;
        if (directory[directory_len - 1] == '/') {
            /* Will insert a '/' separately, so truncate the one provided
               in the directory name. */
            --directory_len;
        }
    } else {
        pathsep_len = 0;
    }
    prefix_len = strlen(prefix);
    suffix_len = strlen(suffix);
    /* sizeof pattern includes its final NULL, so don't add another. */
    len = directory_len  +  pathsep_len  +  prefix_len  +   sizeof pattern  +
        suffix_len;

    if ((template = malloc(len))) {
        char *tp = template;

        (void) strncpy(tp, directory, directory_len);
        tp += directory_len;

        if (pathsep_len == 1) { *tp++ = '/'; }

        (void) strncpy(tp, prefix, prefix_len);
        tp += prefix_len;

        (void) strncpy(tp, pattern, sizeof pattern - 1);
        tp += sizeof pattern - 1;

        (void) strncpy(tp, suffix, suffix_len);
        /* tp += suffix_len; */

        template[len-1] = '\0';

        return template;
    } else {
        perror("malloc");
        return NULL;
    }
}


#if NMH
#include <h/mh.h>

#if HAVE_MKSTEMPS
#   define MHFIXMSG_SWITCHES \
    X("directory", 0, DIRECTORYSW) \
    X("prefix", 0, PREFIXSW) \
    X("suffix", 0, SUFFIXSW) \
    X("version", 0, VERSIONSW) \
    X("help", 0, HELPSW)
#else  /* ! HAVE_MKSTEMPS */
#   define MHFIXMSG_SWITCHES \
    X("directory", 0, DIRECTORYSW) \
    X("prefix", 0, PREFIXSW) \
    X("version", 0, VERSIONSW) \
    X("help", 0, HELPSW)
#endif /* ! HAVE_MKSTEMPS */

#define X(sw, minchars, id) id,
DEFINE_SWITCH_ENUM(MHFIXMSG);
#undef X

#define X(sw, minchars, id) { sw, minchars, id },
DEFINE_SWITCH_ARRAY(MHFIXMSG, switches);
#undef X

void
process_args(int argc, char **argv, const char **directory,
             const char **prefix, const char **suffix) {
    char **argp, **arguments, *cp, buf[100];
#   if ! HAVE_MKSTEMPS
    NMH_UNUSED(suffix);
#   endif /* ! HAVE_MKSTEMPS */

    if (nmh_init(argv[0], 1)) { done(NOTOK); }
    arguments = getarguments (invo_name, argc, argv, 1);
    argp = arguments;

    /*
     * Parse arguments
     */
    while ((cp = *argp++)) {
        if (*cp == '-') {
            switch (smatch(++cp, switches)) {
            case AMBIGSW:
                ambigsw(cp, switches);
                done(NOTOK);
            case UNKWNSW:
                advise (NULL, "-%s unknown", cp);
                (void) snprintf(buf, sizeof buf, "%s [switches]", invo_name);
                print_help(buf, switches, 1);
                done(NOTOK);
            case HELPSW:
                (void) snprintf(buf, sizeof buf, "%s [switches]", invo_name);
                print_help(buf, switches, 1);
                done(OK);
            case VERSIONSW:
                print_version(invo_name);
                done(OK);

            case DIRECTORYSW:
                /* Allow the directory to start with '-'. */
                if ((cp = *argp++) == NULL) {
                    adios(NULL, "missing argument to %s", argp[-2]);
                }
                *directory = cp;
                continue;

            case PREFIXSW:
                /* Allow the prefix to start with '-'. */
                if ((cp = *argp++) == NULL) {
                    adios(NULL, "missing argument to %s", argp[-2]);
                }
                *prefix = cp;
                continue;

#           if HAVE_MKSTEMPS
            case SUFFIXSW:
                /* Allow the suffix to start with '-'. */
                if ((cp = *argp++) == NULL) {
                    adios(NULL, "missing argument to %s", argp[-2]);
                }
                *suffix = cp;
                continue;
#           endif /* HAVE_MKSTEMPS */
            }
        }
    }
}
#else  /* ! NMH */
void
process_args(int argc, char **argv, const char **directory,
             const char **prefix, const char **suffix) {
#   if HAVE_MKSTEMPS
    const char usage[] =
        "usage: %s [-h] [-d directory] [-p prefix] [-s suffix]\n";
    const char optstring[] = "d:hp:s:";
#   else  /* ! HAVE_MKSTEMPS */
    const char usage[] = "usage: %s [-h] [-d directory] [-p prefix]\n";
    const char optstring[] = "d:hp:";
#   endif /* ! HAVE_MKSTEMPS */
    int opt;

    while ((opt = getopt(argc, argv, optstring)) != -1) {
        switch (opt) {
        case 'd':
            *directory = optarg;
            break;
        case 'p':
            *prefix = optarg;
            break;
        case 's':
            *suffix = optarg;
            break;
        case 'h':
            (void) printf(usage, argv[0]);
            exit(0);
        default:
            (void) fprintf(stderr, usage, argv[0]);
            exit(-1);
        }
    }
}
#endif /* ! NMH */
